dev-lang/rust-9999 - cargo clippy: error[E0463]: can't find crate for `core`

Redirected here from https://bugs.gentoo.org/667026

After emerging rust-9999 with clippy useflag on, running "cargo clippy" fails with:

```
$ cargo clippy
    Checking libc v0.2.43                                                                                                                                                         
    Checking unicode-width v0.1.5                                                                                                                                                 
    Checking bitflags v1.0.4                                                                                                                                                      
    Checking cfg-if v0.1.5                                                                                                                                                        
error[E0463]: can't find crate for `core`                                                                                                                                         
                                                                                                                                                                                  
error: aborting due to previous error                                                                                                                                             
                                                                                                                                                                                  
For more information about this error, try `rustc --explain E0463`.                                                                                                               
error: Could not compile `cfg-if`.                                                                                                                                                
warning: build failed, waiting for other jobs to finish...
error[E0463]: can't find crate for `std`                                                                                                                                          
                                                                                                                                                                                  
error: aborting due to previous error                                                                                                                                             
                                                                                                                                                                                  
For more information about this error, try `rustc --explain E0463`.                                                                                                               
error[E0463]: can't find crate for `core`                                                                                                                                         
                                                                                                                                                                                  
error: aborting due to previous error                                                                                                                                             
                                                                                                                                                                                  
For more information about this error, try `rustc --explain E0463`.                                                                                                               
error: Could not compile `libc`.                                                                                                                                                  
warning: build failed, waiting for other jobs to finish...
error: Could not compile `bitflags`.                                                                                                                                              
warning: build failed, waiting for other jobs to finish...
error[E0463]: can't find crate for `core`                                                                                                                                         
                                                                                                                                                                                  
error: aborting due to previous error                                                                                                                                             
                                                                                                                                                                                  
For more information about this error, try `rustc --explain E0463`.                                                                                                               
error: Could not compile `unicode-width`.                                                                                                                                         

To learn more, run the command again with --verbose.
```

--- a/dev-lang/rust/files/rust-9999-fix-clippy-sysroot.patch
+++ b/dev-lang/rust/files/rust-9999-fix-clippy-sysroot.patch
@@ -1,0 +1,64 @@
+diff --git a/src/driver.rs b/src/driver.rs
+index 99f8bc61..02ffc6db 100644
+--- a/src/tools/clippy/src/driver.rs
++++ b/src/tools/clippy/src/driver.rs
+@@ -29,54 +30,22 @@ pub fn main() {
+             exit(0);
+         }
+ 
+-        let sys_root = option_env!("SYSROOT")
+-            .map(String::from)
+-            .or_else(|| std::env::var("SYSROOT").ok())
+-            .or_else(|| {
+-                let home = option_env!("RUSTUP_HOME").or(option_env!("MULTIRUST_HOME"));
+-                let toolchain = option_env!("RUSTUP_TOOLCHAIN").or(option_env!("MULTIRUST_TOOLCHAIN"));
+-                home.and_then(|home| toolchain.map(|toolchain| format!("{}/toolchains/{}", home, toolchain)))
+-            })
+-            .or_else(|| {
+-                Command::new("rustc")
+-                    .arg("--print")
+-                    .arg("sysroot")
+-                    .output()
+-                    .ok()
+-                    .and_then(|out| String::from_utf8(out.stdout).ok())
+-                    .map(|s| s.trim().to_owned())
+-            })
+-            .expect("need to specify SYSROOT env var during clippy compilation, or use rustup or multirust");
+-
+         // Setting RUSTC_WRAPPER causes Cargo to pass 'rustc' as the first argument.
+         // We're invoking the compiler programmatically, so we ignore this/
+-        let mut orig_args: Vec<String> = env::args().collect();
+-        if orig_args.len() <= 1 {
++        let mut args: Vec<String> = env::args().collect();
++        if args.len() <= 1 {
+             std::process::exit(1);
+         }
+-        if Path::new(&orig_args[1]).file_stem() == Some("rustc".as_ref()) {
++        if Path::new(&args[1]).file_stem() == Some("rustc".as_ref()) {
+             // we still want to be able to invoke it normally though
+-            orig_args.remove(1);
++            args.remove(1);
+         }
+-        // this conditional check for the --sysroot flag is there so users can call
+-        // `clippy_driver` directly
+-        // without having to pass --sysroot or anything
+-        let mut args: Vec<String> = if orig_args.iter().any(|s| s == "--sysroot") {
+-            orig_args.clone()
+-        } else {
+-            orig_args
+-                .clone()
+-                .into_iter()
+-                .chain(Some("--sysroot".to_owned()))
+-                .chain(Some(sys_root))
+-                .collect()
+-        };
+ 
+         // this check ensures that dependencies are built but not linted and the final
+         // crate is
+         // linted but not built
+         let clippy_enabled = env::var("CLIPPY_TESTS").ok().map_or(false, |val| val == "true")
+-            || orig_args.iter().any(|s| s == "--emit=dep-info,metadata");
++            || args.iter().any(|s| s == "--emit=dep-info,metadata");
+ 
+         if clippy_enabled {
+             args.extend_from_slice(&["--cfg".to_owned(), r#"feature="cargo-clippy""#.to_owned()]);

--- a/dev-lang/rust/rust-9999.ebuild
+++ b/dev-lang/rust/rust-9999.ebuild
@@ -46,7 +46,7 @@ DEPEND="${RDEPEND}
 	)
 	net-misc/curl
 	cargo? ( !dev-util/cargo )
-    rustfmt? ( !dev-util/rustfmt )
+	rustfmt? ( !dev-util/rustfmt )
 	dev-util/cmake
 	sanitize? ( >=sys-kernel/linux-headers-3.2 )
 	system-llvm? ( >=sys-devel/llvm-3.8.1-r2:=
@@ -55,9 +55,12 @@ DEPEND="${RDEPEND}
 PDEPEND="!cargo? ( >=dev-util/cargo-${CARGO_DEPEND_VERSION} )"
 
 REQUIRED_USE=" || ( ${ALL_LLVM_TARGETS[*]} )
-wasm? ( !system-llvm ) " 
+wasm? ( !system-llvm )
+rls? ( source )"
 
 S="${WORKDIR}/${MY_P}-src"
+
+PATCHES="${FILESDIR}/${P}-fix-clippy-sysroot.patch"
 
 toml_usex() {
 	usex "$1" true false
@@ -109,7 +112,7 @@ src_configure() {
 	fi
 	if use rls; then
 		extended="true"
-		tools="\"rls\",$tools"
+		tools="\"rls\",\"analysis\",$tools"
 	fi
 	if use rustfmt; then
 		extended="true"
@@ -205,14 +208,14 @@ src_configure() {
 
 src_compile() {
 	env $(cat "${S}"/config.env)\
-		${EPYTHON} x.py build --verbose --config="${S}"/config.toml $(echo ${MAKEOPTS} | egrep -o '(\-j|\-\-jobs)(=?|[[:space:]]*)[[:digit:]]+') \
+		"${EPYTHON}" ./x.py build --verbose --config="${S}"/config.toml -j$(makeopts_jobs) \
 		--exclude src/tools/miri || die # https://github.com/rust-lang/rust/issues/52305
 }
 
 src_install() {
 	local rust_target abi_libdir
 
-	env DESTDIR="${D}" ${EPYTHON} x.py install  --verbose --config="${S}"/config.toml $(echo ${MAKEOPTS} | egrep -o '(\-j|\-\-jobs)(=?|[[:space:]]*)[[:digit:]]+') || die
+	env DESTDIR="${D}" "${EPYTHON}" ./x.py install  --verbose --config="${S}"/config.toml -j$(makeopts_jobs) || die
 
 	mv "${D}/usr/bin/rustc" "${D}/usr/bin/rustc-${PV}" || die
 	mv "${D}/usr/bin/rustdoc" "${D}/usr/bin/rustdoc-${PV}" || die

